### PR TITLE
1910-duplicate-trader-payment-requests-vvv-ee

### PR DESCRIPTION
**Pull request by lauragurney**
https://trello.com/c/6i4wVtrb/1910-duplicate-trader-payment-requests-vvv-ee

Code largely lifted from https://github.com/neontribe/ARCVService/compare/1.15/release...1.14/1910-payment-race-conditions

Tested on 44 vouchers in local env

### CODE REVIEW

File: app/Http/Controllers/API/VoucherController.php [hunk 1]
```diff
@@ -126,14 +134,22 @@ public function transition(ApiTransitionVoucherRequest $request): JsonResponse
...
+            if (!empty($vouchers_for_payment)) {
+                \Log::info('SENDING MAIL ' . count($vouchers_for_payment));
+                $this->emailVoucherPaymentRequest($trader, $stateToken, $vouchers_for_payment);
+            }
+            $lock->release();
+            return response()->json(
+                $this->constructResponseMessage($responses)
+            );
+        } else {
+            \Log::info('No lock for me!');
```
Comment (charlesstrange2) [inline app/Http/Controllers/API/VoucherController.php:147]: let's keep this one though...

**Review by charlesstrange2 — CHANGES_REQUESTED** [review]
:wood: :axe:

File: app/Http/Controllers/API/VoucherController.php [hunk 2]
```diff
@@ -24,15 +26,21 @@ class VoucherController extends Controller
...
+        $store = new SemaphoreStore();
+        $factory = new LockFactory($store);
+        $lock = $factory->createLock('transition');
 
-        //create unique, cleaned vouchers
-        $uniqueVouchers = array_unique(
-            Voucher::cleanCodes($request->input('vouchers'))
-        );
+        if ($lock->acquire()) {
+            \Log::info('I have a lock');
```
[the remark this answers is not in the document]
Comment (lauragurney) [reply app/Http/Controllers/API/VoucherController.php]: I chopped it

**Review by charlesstrange2 — APPROVED** [review]
okedoke

**Pull request merged**